controller/controller.go

```diff
@@ -46,29 +48,48 @@ func (c *Controller) Run(ctx context.Context) {
 	}
 }
 
-// reconcile first lists all resources and reconciles them with Controller's state.
+// reconcile first lists all resources and then reconciles them with Controller's state.
 func (c *Controller) reconcile() error {
 	c.Log.Info("starting reconcile")
 	resources, err := c.Resources.List()
 	if err != nil {
 		return fmt.Errorf("listing resources: %w", err)
 	}
-
+	currentResourceIDs := make(map[ResourceID]struct{})
 	for _, resource := range resources {
 		resourceID, err := resource.ID()
 		if err != nil {
 			return fmt.Errorf("getting resource ID: %w", err)
 		}
+
+		currentResourceIDs[resourceID] = struct{}{}
+
 		if _, ok := c.resourceState[resourceID]; !ok {
 			err = resource.Upsert()
 			if err != nil {
-				c.Log.Error("error upserting resource", "err", err)
+				c.Log.Error("error upserting resource", "err", err, "id", resourceID)
 				continue
 			}
-			c.resourceState[resourceID] = struct{}{}
+			c.resourceState[resourceID] = resource
+		}
+	}
+
+	// Prune any resources in the resource state that no longer exist.
+	for id, resource := range c.resourceState {
+		if _, ok := currentResourceIDs[id]; !ok {
+			// Delete the resource.
+			err = resource.Delete()
```

Comment:
Trying to understand when we actually call `resource.Delete()`:

* `c.Resources.List()` returns a list of Tasks (effectively)
* `c.resourceState` is keyed by the TaskDefinitionARN
* ListTasks returns recently stopped tasks [for at least one hour](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ListTasks.html)

So, @ishustava is this a problematic sequence of events, or am I misunderstanding?

* I stop all tasks for TaskDefinitionARN `my-task:1`
* I start tasks for TaskDefinitionARN `my-task:2`
* The controller calls `resource.Upsert()` for `my-task:2`, does nothing (token already exists), and stores `my-task:2` in resourceState
* I wait long enough
* The controller no longer finds tasks for `my-task:1`
* The controller calls `resource.Delete` (for `my-task:1`)
* The controller deletes the ACL token in Consul for `my-task:1` and clears the AWS Secret
* Next iteration, `reconcile` is called. The controller still has `my-task:2` in resourceState, so Upsert is not called and the token is not recreated. Then, I restart tasks which use the `my-task:2` TaskDefinitionARN, which can't start without their ACL token.

Reply:
Great question! 

I don't think this will be problematic. When you stop a task, you're effectively changing its desired status from `RUNNING` to `STOPPED`. By default, `ListTasks` will return tasks with desired status of `RUNNING`, and so listing tasks will not return the stopped tasks from step 1.

Reply:
good question, I'll try to reproduce, although off the top of my head I'm not sure how it could be possible to start two tasks with different task definition versions. I'll try to test out different scaling scenarios and see if I run into any edge cases. 

Thanks for thinking this through ❤️

Reply:
> although off the top of my head I'm not sure how it could be possible to start two tasks with different task definition versions

An ECS Service does a rolling update by default (I think). So if you have multiple tasks per service, you could have two task definition versions running during a deployment.

[Docs](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/update-service.html):

> If a service is using the rolling update (ECS) deployment type, the minimum healthy percent represents a lower limit on the number of tasks in a service that must remain in the RUNNING state during a deployment, as a percentage of the desired number of tasks (rounded up to the nearest integer). ... For example, if your service has a desired number of four tasks and a minimum healthy percent of 50%, the scheduler may stop two existing tasks to free up cluster capacity before starting two new tasks.

The other way is to not use services. Just call `RunTask` and pass it the task definition versions.

Reply:
Great catch!!! I tested it with service updates, and it does delete the token like you described. I've updated the code to skip token deletion if there's a service registered with Consul. I haven't found that there will be a race condition when you actually delete the service in my testing, but if I'm missing something this is still safer since it's better to leave tokens around then to delete and cause service to not be able to restart.